requestparsing/test_harness.go

```diff
@@ -0,0 +1,120 @@
+package requestparsing
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+)
+
+type fakeHandler struct {
+	// callback is an anonymous function that verifies properties of the HTTP
+	// request and response provided in the test cases.
+	callback func(*http.Request)
+}
+
+func (h *fakeHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.callback(r)
+	io.WriteString(w, "Hello world!")
+}
+
+type fakeListener struct {
+	closeOnce      sync.Once
+	connChannel    chan net.Conn
+	serverEndpoint io.Closer
+	clientEndpoint net.Conn
+}
+
+// newFakeListener creates an instance of fakeListener. This will pass requests
+// to the HTTP server as part of the testing harness.
+func newFakeListener() *fakeListener {
+	s2c, c2s := net.Pipe()
+	c := make(chan net.Conn, 1)
+	c <- s2c
+	return &fakeListener{
+		connChannel:    c,
+		serverEndpoint: s2c,
+		clientEndpoint: c2s,
+	}
+}
+
+// Accept passes a network connection to the HTTP server to
+// enable bidirectional communication with the client. It will return an error
+// if Accept is called after the listener was closed
```

Comment:
nit: here and every godoc comment: make them full sentences with correct punctuation (i.e. dot at the end).